[css-counter-styles] Support CJK counter styles outside of range

igower opened the pull request:
#### bf482c9f31efb9850d69713212428be239fe8727
<pre>
[css-counter-styles] Support CJK counter styles outside of range
<a href="https://bugs.webkit.org/show_bug.cgi?id=257369">https://bugs.webkit.org/show_bug.cgi?id=257369</a>
<a href="https://rdar.apple.com/problem/109875198">rdar://problem/109875198</a>

Reviewed by NOBODY (OOPS!).

Updated the counterForSystemCJK function to account for values in an expanded range. Now in range 10^-16 + 1 to 10^16 -1 for Chinese, Japanese, and Korean styles. For tests, switched out expected behavior to behavior defined in alt-ref files to match increased range.

* LayoutTests/TestExpectations:
* LayoutTests/fast/lists/li-values-expected.txt:
* LayoutTests/fast/lists/li-values.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/counter-japanese-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/counter-japanese-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/css3-counter-styles-049-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/counter-japanese-informal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/counter-japanese-informal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/css3-counter-styles-044-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/counter-korean-hangul-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/counter-korean-hangul-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/css3-counter-styles-054-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/counter-korean-hanja-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/counter-korean-hanja-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/css3-counter-styles-064-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/counter-korean-hanja-informal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/counter-korean-hanja-informal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/css3-counter-styles-059-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/counter-simp-chinese-formal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/counter-simp-chinese-formal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/css3-counter-styles-078-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/counter-simp-chinese-informal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/counter-simp-chinese-informal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/css3-counter-styles-073-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/counter-trad-chinese-formal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/counter-trad-chinese-formal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/css3-counter-styles-088-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/counter-trad-chinese-informal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/counter-trad-chinese-informal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/css3-counter-styles-083-expected.html:
* Source/WTF/wtf/SaturatedArithmetic.h:
(WTF::signedAddInt64Overflows):
(WTF::saturatedSum&lt;int64_t&gt;):
(WTF::saturatedSum):
* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::counterForSystemCJK):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemJapaneseInformal):
(WebCore::CSSCounterStyle::counterForSystemJapaneseFormal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHangulFormal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHanjaInformal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHanjaFormal):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::shouldApplyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::rangeFromCSSValue):
(WebCore::extractSystemDataFromCSSValue):
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
(WebCore::CSSCounterStyleDescriptors::rangesCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style simp-chinese-informal):
(@counter-style simp-chinese-formal):
(@counter-style trad-chinese-informal):
(@counter-style trad-chinese-formal):
(@counter-style japanese-informal):
(@counter-style japanese-formal):
(@counter-style korean-hangul-formal):
(@counter-style korean-hanja-informal):
(@counter-style korean-hanja-formal):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::CounterNode::CounterNode):
(WebCore::CounterNode::create):
(WebCore::CounterNode::computeCountInParent const):
(WebCore::CounterNode::recount):
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/CounterNode.h:
(WebCore::CounterNode::value const):
(WebCore::CounterNode::countInParent const):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::planCounter):
(WebCore::RenderCounter::originalText const):
* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::updateValueNow const):
* Source/WebCore/rendering/style/CounterDirectives.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueCounter):
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/bf482c9f31efb9850d69713212428be239fe8727

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117682 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37360 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/27988 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/123816 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69694 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119560 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38052 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/45942 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89317 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/51086 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120634 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30294 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105519 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/69808 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29357 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23635 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67474 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/109792 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99711 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/23815 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/126909 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/116191 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44585 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33555 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/97974 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/44943 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/101746 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97761 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43140 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21102 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/40950 "The change is no longer eligible for processing.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44457 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50131 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/144889 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/43915 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/37283 "Found 1 new JSC binary failure: testapi, Found 20046 new JSC stress test failures: ChakraCore.yaml/ChakraCore/test/Array/array_slice.js.default, ChakraCore.yaml/ChakraCore/test/Array/array_sort2.js.default, ChakraCore.yaml/ChakraCore/test/Array/nativearray_gen6.js.default, ChakraCore.yaml/ChakraCore/test/Array/reverse1.js.default, ChakraCore.yaml/ChakraCore/test/Array/toLocaleString.js.default, ChakraCore.yaml/ChakraCore/test/ControlFlow/forInPrimitive.js.default, ChakraCore.yaml/ChakraCore/test/Date/date_cache_bug.js.default, ChakraCore.yaml/ChakraCore/test/Error/CallNonFunction.js.default, ChakraCore.yaml/ChakraCore/test/Error/stack2.js.default, ChakraCore.yaml/ChakraCore/test/Function/FuncBody.bug227901.js.default ... (failure)") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47262 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45606 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fantasai commented on LayoutTests/fast/lists/li-values.html:
```diff
@@ -40,10 +40,10 @@
+    <li value="10000" style="list-style-type: japanese-informal;">The list marker should show (decimal 10000) (japanese-informal: -9999999999999999 9999999999999999).</li>
+    <li value="-10000" style="list-style-type: japanese-informal;">The list marker should show (decimal -10000) (japanese-informal: -9999999999999999 9999999999999999).</li>
+    <li value="10000" style="list-style-type: japanese-formal;">The list marker should show (decimal 10000) (japanese-formal: -9999999999999999 9999999999999999).</li>
+    <li value="-10000" style="list-style-type: japanese-formal;">The list marker should show (decimal -10000) (japanese-formal: -9999999999999999 9999999999999999).</li>
```
Might want to use a dash between the numbers, to be consistent with the other entries in this test. :)

fantasai commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -174,46 +175,88 @@ static String counterForSystemCJK(int number, const std::array<UChar, 17>& table
+        }
+
+        addSpace = false;
+        bool trailingZero = true;
```
Maybe add a comment about what these are about?

fantasai commented on LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/css3-counter-styles-088-expected.html:
```diff
@@ -1,13 +1,12 @@
+<link rel="author" title="Mats Palmgren" href="mailto:mats@mozilla.com">
 <link rel='help' href='http://www.w3.org/TR/css-counter-styles-3/#complex-cjk'>
-<link rel="reviewer" title="Chris Lilley" href="mailto:chris@w3.org" />
-<meta name="assert" content="[Exploratory] list-style-type: trad-chinese-formal produces counter values outside its range without using the prescribed fallback style.">
-<style type='text/css'>
+<link rel='help' href='https://bugzilla.mozilla.org/show_bug.cgi?id=1738356'>
+<style>
```
I'm guessing these are import changes? It seems weird to change the credits given the the test writer.

fantasai reviewed (commented):
Do we have test coverage for values such as:

1,2345,6789 (all digits non-zero)
12,0045,3001 ( leading zero in group, sandwiched zeros in group)
987,000,0123 (skipped group, leading zeros and ones)
509,000,1234 (skipped group, leading one)

?

nt1m reviewed (commented):
Could we do the WPT sync separately?

igower committed:
[css-counter-styles] Support CJK counter styles outside of range
https://bugs.webkit.org/show_bug.cgi?id=257369
rdar://problem/109875198

Reviewed by NOBODY (OOPS!).

Updated the counterForSystemCJK function to account for values in an expanded range. Now in range 10^-16 + 1 to 10^16 -1 for Chinese, Japanese, and Korean styles. For tests, switched out expected behavior to behavior defined in alt-ref files to match increased range.

* LayoutTests/TestExpectations:
* LayoutTests/fast/lists/li-values-expected.txt:
* LayoutTests/fast/lists/li-values.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/counter-japanese-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/counter-japanese-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-formal/css3-counter-styles-049-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/counter-japanese-informal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/counter-japanese-informal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/japanese-informal/css3-counter-styles-044-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/counter-korean-hangul-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/counter-korean-hangul-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hangul-formal/css3-counter-styles-054-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/counter-korean-hanja-formal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/counter-korean-hanja-formal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-formal/css3-counter-styles-064-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/counter-korean-hanja-informal-extended-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/counter-korean-hanja-informal-extended-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/korean-hanja-informal/css3-counter-styles-059-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/counter-simp-chinese-formal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/counter-simp-chinese-formal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-formal/css3-counter-styles-078-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/counter-simp-chinese-informal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/counter-simp-chinese-informal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/simp-chinese-informal/css3-counter-styles-073-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/counter-trad-chinese-formal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/counter-trad-chinese-formal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-formal/css3-counter-styles-088-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/counter-trad-chinese-informal-expected.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/counter-trad-chinese-informal-ref.html:
* LayoutTests/imported/w3c/web-platform-tests/css/css-counter-styles/trad-chinese-informal/css3-counter-styles-083-expected.html:
* Source/WTF/wtf/SaturatedArithmetic.h:
(WTF::signedAddInt64Overflows):
(WTF::saturatedSum<int64_t>):
(WTF::saturatedSum):
* Source/WebCore/css/CSSCounterStyle.cpp:
(WebCore::counterForSystemCJK):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemSimplifiedChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseInformal):
(WebCore::CSSCounterStyle::counterForSystemTraditionalChineseFormal):
(WebCore::CSSCounterStyle::counterForSystemJapaneseInformal):
(WebCore::CSSCounterStyle::counterForSystemJapaneseFormal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHangulFormal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHanjaInformal):
(WebCore::CSSCounterStyle::counterForSystemKoreanHanjaFormal):
(WebCore::CSSCounterStyle::initialRepresentation const):
(WebCore::CSSCounterStyle::fallbackText):
(WebCore::CSSCounterStyle::text):
(WebCore::CSSCounterStyle::shouldApplyNegativeSymbols const):
(WebCore::CSSCounterStyle::applyPadSymbols const):
(WebCore::CSSCounterStyle::isInRange const):
* Source/WebCore/css/CSSCounterStyle.h:
* Source/WebCore/css/CSSCounterStyleDescriptors.cpp:
(WebCore::rangeFromCSSValue):
(WebCore::extractSystemDataFromCSSValue):
(WebCore::CSSCounterStyleDescriptors::areSymbolsValidForSystem):
(WebCore::CSSCounterStyleDescriptors::systemCSSText const):
(WebCore::CSSCounterStyleDescriptors::rangesCSSText const):
* Source/WebCore/css/CSSCounterStyleDescriptors.h:
* Source/WebCore/css/CSSCounterStyleRule.cpp:
(WebCore::toCounterStyleSystemEnum):
* Source/WebCore/css/CSSValueKeywords.in:
* Source/WebCore/css/counterStyles.css:
(@counter-style simp-chinese-informal):
(@counter-style simp-chinese-formal):
(@counter-style trad-chinese-informal):
(@counter-style trad-chinese-formal):
(@counter-style japanese-informal):
(@counter-style japanese-formal):
(@counter-style korean-hangul-formal):
(@counter-style korean-hanja-informal):
(@counter-style korean-hanja-formal):
* Source/WebCore/css/parser/CSSPropertyParserConsumer+CounterStyles.cpp:
(WebCore::CSSPropertyParserHelpers::consumeCounterStyleSystem):
* Source/WebCore/rendering/CounterNode.cpp:
(WebCore::CounterNode::CounterNode):
(WebCore::CounterNode::create):
(WebCore::CounterNode::computeCountInParent const):
(WebCore::CounterNode::recount):
(WebCore::showTreeAndMark):
* Source/WebCore/rendering/CounterNode.h:
(WebCore::CounterNode::value const):
(WebCore::CounterNode::countInParent const):
* Source/WebCore/rendering/RenderCounter.cpp:
(WebCore::planCounter):
(WebCore::RenderCounter::originalText const):
* Source/WebCore/rendering/RenderListItem.cpp:
(WebCore::RenderListItem::updateValueNow const):
* Source/WebCore/rendering/style/CounterDirectives.h:
* Source/WebCore/style/StyleBuilderCustom.h:
(WebCore::Style::BuilderCustom::applyValueCounter):

igower commented on LayoutTests/TestExpectations:
```diff
@@ -6614,8 +6614,12 @@ webkit.org/b/297734 imported/w3c/web-platform-tests/css/css-counter-styles/count
 
 imported/w3c/web-platform-tests/css/css-counter-styles/counter-suffix.html [ ImageOnlyFailure ]
 imported/w3c/web-platform-tests/css/css-counter-styles/hebrew/counter-hebrew-nested.html [ ImageOnlyFailure ]
+imported/w3c/web-platform-tests/css/css-lists/counter-reset-increment-overflow-underflow.html [ ImageOnlyFailure ]
```
This test now fails because of the expanded range functionality. Instead of preventing a list from incrementing out of range and defaulting back to the original value, the value is increased but capped at INT_MAX.

darinadler replied:
Could we fix this in the same pull request?

darinadler commented on Source/WTF/wtf/SaturatedArithmetic.h:
```diff
@@ -43,6 +43,53 @@ template<typename SignedIntegralType> std::enable_if_t<std::is_integral_v<Signed
+{
+#if COMPILER_HAS_CLANG_BUILTIN(__builtin_saddll_overflow) && !(defined __clang_major__ && __clang_major__ < 7)
+    // Handle different int64_t definitions across platforms
+#if defined(__APPLE__) || (defined(__SIZEOF_LONG_LONG__) && __SIZEOF_LONG_LONG__ == 8 && !defined(__LP64__))
```
Better to use `if constexpr (sizeof(long long) == 8)` instead of this `__APPLE__` and other macro stuff. Unless there is a problem compiling that?

But also why can't all platforms use `long long`? Is there a platform where `long long` is >64 bits?

darinadler commented on Source/WTF/wtf/SaturatedArithmetic.h:
```diff
@@ -114,7 +161,12 @@ template<typename UnsignedIntegralType> constexpr std::enable_if_t<std::is_integ
 }
 
-template<typename IntegralType, typename... ArgumentTypes> constexpr uint32_t saturatedSum(IntegralType value, ArgumentTypes... arguments)
+template<typename IntegralType> constexpr IntegralType saturatedSum(IntegralType value)
+{
+    return value;
+}
```
Why is this addition needed?

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -224,22 +268,42 @@ static String counterForSystemCJK(int number, const std::array<char16_t, 17>& ta
+    if (needsNegativeSign) {
+        if (cjk == CJK::Japanese) {
+            // Japanese negative sign is a string of 4 characters.
+            characters[length++] = 0x30DE;
+            characters[length++] = 0x30A4;
+            characters[length++] = 0x30CA;
+            characters[length++] = 0x30B9;
```
I don't see the code change making sure we have space for the 3 additional characters, the code above says `bufferLength + 1` and I think the 1 is for the possible negative sign. So it would need to be `bufferLength + 4`.

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -224,22 +268,42 @@ static String counterForSystemCJK(int number, const std::array<char16_t, 17>& ta
             last = character;
+            if (cjk == CJK::Chinese && last == FourthGroupMarker && table[character - 1] == 0x4E07) {
+                // Simple Chinese fourth group marker is 2 characters.
+                characters[length++] = 0x4EBF;
```
What makes sure we have space for this in the buffer?

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
 {
-    static constexpr std::array<char16_t, 17> simplifiedChineseInformalTable {
-        0x842C, 0x5104, 0x5146, // These three group markers are probably wrong; OK because we don't use this on big enough numbers.
+    static constexpr std::array<UChar, 18> simplifiedChineseInformalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> simplifiedChineseInformalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
 {
-    static constexpr std::array<char16_t, 17> simplifiedChineseFormalTable {
-        0x842C, 0x5104, 0x5146, // These three group markers are probably wrong; OK because we don't use this on big enough numbers.
+    static constexpr std::array<UChar, 18> simplifiedChineseFormalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> simplifiedChineseFormalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
 {
-    static constexpr std::array<char16_t, 17> traditionalChineseFormalTable {
-        0x842C, 0x5104, 0x5146, // These three group markers are probably wrong; OK because we don't use this on big enough numbers.
+    static constexpr std::array<UChar, 18> traditionalChineseFormalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> traditionalChineseFormalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
+
+String CSSCounterStyle::counterForSystemJapaneseInformal(int64_t value)
+{
+    static constexpr std::array<UChar, 18> japaneseInformalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> japaneseInformalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
+
+String CSSCounterStyle::counterForSystemJapaneseFormal(int64_t value)
+{
+    static constexpr std::array<UChar, 18> japaneseFormalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> japaneseFormalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -353,9 +477,11 @@ String CSSCounterStyle::counterForSystemEthiopicNumeric(unsigned value)
+String CSSCounterStyle::initialRepresentation(int64_t value, WritingMode writingMode) const
 {
     unsigned absoluteValue = std::abs(value);
+    if (abs(value) > INT_MAX)
+        absoluteValue = value < 0 ? INT_MIN: INT_MAX;
```
What's the rationale here? This is very subtle code using the non-overloaded version of `abs` and using the 32 bit range on a 64 bit integer so it needs an explanation and test covering it. I'm assuming there is a good reason for this.

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
+
+String CSSCounterStyle::counterForSystemKoreanHangulFormal(int64_t value)
+{
+    static constexpr std::array<UChar, 18> koreanHangulFormalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> koreanHangulFormalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
+
+String CSSCounterStyle::counterForSystemKoreanHanjaInformal(int64_t value)
+{
+    static constexpr std::array<UChar, 18> koreanHanjaInformalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> koreanHanjaInformalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -263,52 +327,112 @@ String CSSCounterStyle::counterForSystemDisclosureOpen(WritingMode writingMode)
+
+String CSSCounterStyle::counterForSystemKoreanHanjaFormal(int64_t value)
+{
+    static constexpr std::array<UChar, 18> koreanHanjaFormalTable {
```
```suggestion
    static constexpr std::array<char16_t, 18> koreanHanjaFormalTable {
```

darinadler commented on Source/WebCore/css/CSSCounterStyle.cpp:
```diff
@@ -465,7 +601,12 @@ bool CSSCounterStyle::isInRange(int value) const
+        case CSSCounterStyleDescriptors::System::KoreanHangulFormal:
+        case CSSCounterStyleDescriptors::System::KoreanHanjaInformal:
+        case CSSCounterStyleDescriptors::System::KoreanHanjaFormal:
+            return true;
```
Should not be a second group of styles that return true. These styles should move into the group at the top of the `switch`.

darinadler commented on Source/WebCore/css/counterStyles.css:
```diff
@@ -302,94 +302,85 @@
 @counter-style simp-chinese-informal {
     system: -internal-simplified-chinese-informal;
-    range: -9999 9999;
+    range: calc(-1 * pow(10, 16) + 1) calc(pow(10, 16) - 1);
```
Should we write the number out instead of calculating it? Maybe a performance cost to parse and calculate each time.

darinadler commented on Source/WebCore/rendering/CounterNode.cpp:
```diff
@@ -350,7 +350,7 @@ static void showTreeAndMark(const CounterNode* node)
         for (const CounterNode* parent = current; parent && parent != root; parent = parent->parent())
             fprintf(stderr, "    ");
-        SAFE_FPRINTF(stderr, "%p %s: %d %d P:%p PS:%p NS:%p R:%p\n",
+        SAFE_FPRINTF(stderr, "%p %s: %lld %lld P:%p PS:%p NS:%p R:%p\n",
```
Either need to cast to `long long` or use the pepper format string for int64_t, something like `PRI64`, although I didn't look it up. This incorrectly uses the format string for `long long` but that isn't guaranteed to be the same, right?

darinadler reviewed (changes requested):
The most important part of this to get right is to make sure we don't introduce some new security problems with buffer overruns and I'm pretty sure there are some mistakes that do just that that I commented about below. This set of enhancements and fixes isn't worth introducing a security vulnerability so we should make sure to get that part right above all else.

I think there are not enough new tests here. While the comment talks only about extending the range of CJK, there are all sorts of newly supported values and edge cases. I don't see tests about super huge counter values and the highest values and exactly what happens when you run over the edge. Maybe I've overlooked them?

Generally speaking, switching to 64-bit must have some value for the non-CJK numbering systems too. Where are the tests showing that working?
